yt_dlp/extractor/crunchyroll.py

```diff
@@ -47,52 +52,59 @@ def _perform_login(self, username, password):
-                    'Authorization': 'Basic ' + str(base64.b64encode(('%s:' % app_config['cxApiParams'][key]).encode('ascii')), 'ascii')
-                }, data=f'grant_type={grant_type}'.encode('ascii'))
-            policy_response = self._download_json(
-                f'{api_domain}/index/v2', None, note='Retrieving signed policy',
-                headers={
-                    'Authorization': auth_response['token_type'] + ' ' + auth_response['access_token']
-                })
-            cms = policy_response.get('cms_web')
-            bucket = cms['bucket']
-            params = {
-                'Policy': cms['policy'],
-                'Signature': cms['signature'],
-                'Key-Pair-Id': cms['key_pair_id']
-            }
-            locale = traverse_obj(initial_state, ('localization', 'locale'))
-            if locale:
-                params['locale'] = locale
-            CrunchyrollBaseIE.params = (api_domain, bucket, params)
-        return CrunchyrollBaseIE.params
+    def _update_query(self, lang):
+        if not self._QUERY:
+            self._QUERY = {}
+
+        if lang in self._QUERY:
+            return
+
+        webpage = self._download_webpage(
+            f'{self._BASE_URL}/{lang}', None, note=f'Retrieving main page (lang={lang or None})')
+
+        initial_state = self._search_json(r'__INITIAL_STATE__\s*=', webpage, 'initial state', None)
+        self._QUERY[lang] = traverse_obj(initial_state, {
+            'locale': ('localization', 'locale'),
+        }) or None
+
+        if self._BASIC_AUTH:
+            return
+
+        app_config = self._search_json(r'__APP_CONFIG__\s*=', webpage, 'app config', None)
+        cx_api_param = app_config['cxApiParams']['accountAuthClientId' if self.is_logged_in else 'anonClientId']
+        self.write_debug(f'Using cxApiParam={cx_api_param}')
+        self._BASIC_AUTH = 'Basic ' + base64.b64encode(f'{cx_api_param}:'.encode()).decode()
+
+    def _update_auth(self):
+        if self._AUTH_HEADERS and self._AUTH_REFRESH > time_seconds():
+            return
+
+        assert self._BASIC_AUTH, '_update_query needs to be called at least one time beforehand'
+        grant_type = 'etp_rt_cookie' if self.is_logged_in else 'client_id'
+        auth_response = self._download_json(
+            f'{self._BASE_URL}/auth/v1/token', None, note=f'Authenticating with grant_type={grant_type}',
+            headers={'Authorization': self._BASIC_AUTH}, data=f'grant_type={grant_type}'.encode())
+
+        self._AUTH_HEADERS = {'Authorization': auth_response['token_type'] + ' ' + auth_response['access_token']}
+        self._AUTH_REFRESH = time_seconds(seconds=traverse_obj(auth_response, ('expires_in', {float_or_none}), default=300))
+
+    def _call_api(self, endpoint, internal_id, lang=None, note='api'):
+        self._update_query(lang)
+        self._update_auth()
+
+        endpoint = remove_start(endpoint, 'cms:/')
+        if 'cms' not in endpoint:
```

Comment:
This feels a bit dangerous. The endpoints don't currently contain the letters `cms` in order anywhere afaik, and the user supplied data uses capital letters, but still, this seems like it's just asking for a weird bug later.

I think it should be fine to test for a prefix of `content/v2/cms/` here, as if that prefix changes, the extractor would break anyway.

Reply:
`.startswith()` will be more solid, I agree